src/data/finset/sort.lean

```diff
@@ -52,28 +52,31 @@ theorem sort_sorted_lt (s : finset α) :
   list.sorted (<) (sort (≤) s) :=
 (sort_sorted _ _).imp₂ (@lt_of_le_of_ne _ _) (sort_nodup _ _)
 
-lemma sorted_zero_eq_min' (s : finset α) (h : 0 < (s.sort (≤)).length) (H : s.nonempty) :
-  (s.sort (≤)).nth_le 0 h = s.min' H :=
+lemma sorted_zero_eq_min' (s : finset α) (h : 0 < s.card) :
+  (s.sort (≤)).nth_le 0 (by rwa length_sort) = s.min' (by rwa card_pos at h) :=
```

Comment:
This might just be my inexperience, but are lemmas that use tactics in their statements easy to use?

Reply:
the presence of tactics is not the problem so much as the choice of a particular proof to go in that slot. Most tactics will ignore the hypothesis argument when matching up to defeq but occasionally it helps to have a free variable for the hypothesis. Having free variables on both sides of a rewrite can be a problem, though, since it doesn't actually pin down what the hypothesis in the RHS is, and so you get an extra subgoal you didn't want. So it's sometimes helpful to have both versions, or a version where the LHS has a variable and the RHS is concrete (if it is to be used for left to right rewrites or simps).

Reply:
@digama0 So, what do you suggest? Should we revert, or continue?

Reply:
Since the tactic is used to prove a Prop, it will match any Prop that is given to the lemma if you rewrite from left to right or from right to left, by proof irrelevance, and it will provide the required prop on the output of the rewrite. So, to me, this is an improvement over the previous situation, where the user had to provide more data. In all uses of this lemma, the outcome of the change is that we can remove one of the arguments to the call.

Mario, you say that occasionally it helps to have a free variable for the hypothesis. Do you have an example of this behavior?

Reply:
I agree that the original form with two variables is not good for anyone, except in the case where you are matching an equality in the goal. But the new version of the lemma can *never* prove `h` by unification, while the original could, which means you are always going to have to prove the hypothesis `h` as a subgoal when you apply the new lemma.

Is there a definite common rewrite direction? If so, (say l-t-r) you can put the free hypothesis on the LHS (as is), and prove the RHS hypothesis using two rewrites on the LHS hypothesis. That way if you rewrite the hypothesis will be proven by unification, and you don't get any subgoal for the RHS because it's been proven in the theorem. 0 subgoals is better than 1, no?

If you don't know how the lemma will be used, I suggest providing both the original over-general theorem (this one getting a prime) and the present constrained theorem, and then users can use an appropriate version for the application. There are a few list lemmas like this.

Reply:
I think only a couple of these are used, and then only left-to-right. But if you are already considering two versions of each, then maybe one could work left-to-right and one right-to-left? Something like this:

`lemma sorted_zero_eq_inf' (s : finset α) (h : 0 < (s.sort (≤)).length) :
  (s.sort (≤)).nth_le 0 h = s.1.inf' (by rwa [length_sort, card_pos] at h) :=
begin ... end`

`lemma inf'_eq_sorted_zero (s : finset α) (H : s.nonempty) :
  s.1.inf' H = (s.sort (≤)).nth_le 0 (by rwa [length_sort, card_pos]) :=
by rw sorted_zero_eq_inf'`

`lemma sorted_last_eq_sup' (s : finset α) (h : (s.sort (≤)).length - 1 < (s.sort (≤)).length) :
  (s.sort (≤)).nth_le ((s.sort (≤)).length - 1) h =
  s.1.sup' (by simpa [card_pos] using lt_of_le_of_lt (nat.zero_le _) h) :=
begin ... end`

`lemma sup'_eq_sorted_last (s : finset α) (H : s.nonempty) :
  s.1.sup' H = (s.sort (≤)).nth_le ((s.sort (≤)).length - 1)
                                   (by simpa using sub_lt (card_pos.mpr H) zero_lt_one) :=
by rw sorted_last_eq_sup'`

`lemma mono_of_fin_zero_eq_inf' {s : finset α} {k : ℕ} (h : s.card = k) (hz : 0 < k) :
  mono_of_fin s h ⟨0, hz⟩ = s.1.inf' (card_pos.1 (h.symm ▸ hz)) :=
begin ... end`

`lemma inf'_eq_mono_of_fin_zero {s : finset α} {k : ℕ} (h : s.card = k) (hs : s.nonempty) (hz : 0 < k) :
  s.1.inf' hs = mono_of_fin s rfl ⟨0, card_pos.2 hs⟩ :=
by rw mono_of_fin_zero`

`lemma mono_of_fin_last_eq_sup' {s : finset α} {k : ℕ} (h : s.card = k) (hz : 0 < k) :
  mono_of_fin s h ⟨k-1, sub_lt hz zero_lt_one⟩ = s.1.sup' (card_pos.1 (h.symm ▸ hz)) :=
begin ... end`

`lemma sup'_eq_mono_of_fin_last {s : finset α} (hs : s.nonempty) :
  s.1.sup' hs = mono_of_fin s rfl ⟨s.card - 1, sub_lt (card_pos.2 hs) zero_lt_one⟩ :=
by rw mono_of_fin_last rfl (card_pos.2 hs)`

`lemma mono_of_fin_singleton_eq (a : α) (i : fin 1) {h} :
  mono_of_fin {a} h i = a :=
begin ... end`

`lemma eq_mono_of_fin_singleton (a : α) (i : fin 1) :
  a = mono_of_fin {a} (card_singleton a) i :=
by rw mono_of_fin_singleton`